bodo/numba_compat.py

```diff
@@ -6847,3 +6848,65 @@ def _from_python_function(cls, func_ir, typemap, restype,
 
 
 numba.core.funcdesc.FunctionDescriptor._from_python_function = _from_python_function
+
+
+class BodoCacheLocator(numba.core.caching._CacheLocator):
+    """
+    A CacheLocator for Numba that handles functions created from strings.
+    """
+    __slots__ = ('_py_file', '_cache_path', '_bytes_source')
+    registered_funcs = {}   # Holds mapping of generated function name to its source.
+    if os.environ.get("BODO_PLATFORM_CACHE_LOCATION") is not None:
+        cache_path = numba.config.CACHE_DIR
+    else:
+        appdirs = AppDirs(appname="bodo", appauthor=False)
+        cache_path = os.path.join(appdirs.user_cache_dir, ".bodo_strfunc_cache")
+
+    def __init__(self, py_func, py_file):
+        source = BodoCacheLocator.registered_funcs[py_func.__qualname__]
+        if isinstance(source, bytes):
+            self._bytes_source = source
+        else:
+            self._bytes_source = source.encode('utf-8')
+
+    def get_cache_path(self):
+        return BodoCacheLocator.cache_path
+
+    def get_source_stamp(self):
+        return hashlib.sha256(self._bytes_source).hexdigest()
+
+    def get_disambiguator(self):
+        firstlines = b''.join(self._bytes_source.splitlines(True)[:3])
+        return hashlib.sha256(firstlines).hexdigest()[:10]
+
+    @classmethod
+    def from_function(cls, py_func, py_file):
+        if not py_func.__qualname__.startswith("bodo"):
+            return
+        self = cls(py_func, py_file)
+        try:
+            self.ensure_cache_path()
+        except OSError:
+            # Cannot ensure the cache directory exists
+            return
+        return self
+
+    @classmethod
+    def register(cls, name, source):
```

Comment:
Where is this called? I can't find it in Numba source code.

Reply:
Called by bodo/utils/utils/bodo_exec, not Numba.